cmd/zoekt-sourcegraph-indexserver/queue.go

```diff
@@ -128,6 +136,96 @@ func (q *Queue) Iterate(f func(*IndexOptions)) {
 	}
 }
 
+// debugIteratedOrdered will call f on each queueItem (sorted by indexing priority)
+// known to the queue, including queueItems that have been popped from the queue).
+//
+// Note: The mutex is held during all calls to f. Callers must not modify the queueItems.
+func (q *Queue) debugIteratedOrdered(f func(*queueItem)) {
+	q.mu.Lock()
+	defer q.mu.Unlock()
+
+	queueItems := make([]*queueItem, 0, len(q.items))
+	for _, item := range q.items {
+		queueItems = append(queueItems, item)
+	}
+
+	sort.Slice(queueItems, func(i, j int) bool {
+		x, y := queueItems[i], queueItems[j]
+
+		if x.heapIdx < 0 && y.heapIdx >= 0 {
+			// if x is popped, but y isn't - then put x at the back
+			return false
+		}
+
+		if y.heapIdx < 0 && x.heapIdx >= 0 {
+			// if y is popped, but x isn't - then ensure x is at the front
+			return true
+		}
+
+		return lessQueueItemPriority(x, y)
+	})
+
+	for _, item := range queueItems {
+		f(item)
+	}
+}
+
+func (q *Queue) handleDebugQueue(w http.ResponseWriter, r *http.Request) {
+	if r.Method != http.MethodGet {
+		http.Error(w, "method must be GET", http.StatusBadRequest)
+		return
+	}
+
+	w.Header().Set("Content-Type", "text/plain")
+
+	printHeaders := true
+
+	if raw := r.URL.Query().Get("header"); raw != "" {
+		parsed, err := strconv.ParseBool(raw)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("invalid %q parameter: %s", "header", err), http.StatusBadRequest)
+			return
+		}
+
+		printHeaders = parsed
+	}
+
+	writer := tabwriter.NewWriter(w, 16, 8, 4, ' ', 0)
```

Comment:
minor: instead of writing directly to `w`, can you write to an intermediate `bytes.Buffer`. Unlikely a problem given this is an internal debug command, but holding a coarse lock while a writing to a user leads to slow users blocking up programs (or slowlaris attacks on public endpoints). I don't think we need to be concerned of memory use in this case.